ngcc: re-export private NgModule declarations

petebacondarwin opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mary-poppins commented:
You can preview fd6cbc6 at https://pr26906-fd6cbc6.ngbuilds.io/.

mary-poppins commented:
You can preview 1ae87b6 at https://pr26906-1ae87b6.ngbuilds.io/.

Toxicable commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -121,7 +120,13 @@ export class DecorationAnalyzer {
     });
-    return {constantPool, sourceFile: analyzedFile.sourceFile, compiledClasses};
+    const declarations: R3Reference[] = [];
+    compiledClasses.forEach(c => {
```
Looks like you've got `forEach` a few times here, unless im missing something they should be replace with for loops.
ref: https://github.com/angular/angular/blob/master/packages/core/src/render3/PERF_NOTES.md#loops

JoostK replied:
@Toxicable those rules are mostly only adhered to in the runtime implementation, not in the compiler infrastructure.

Toxicable replied:
I think it would be a good idea to adheard to in throughout the codebase.
But I see your point that changing it wouldn't produce any noticable impact to users.

petebacondarwin replied:
Interesting point @Toxicable. Performance is always a trade-off. In this case this function, `getPrivateDeclarations()` only gets called once per entry-point and also `rootFiles` should be very small (hopefully just 1 - although there might be stdlibs, etc).

Given the amount of calls to things like `map()` and `filter()` I doubt that optimising this particular bit of the code will make much difference.

Finally, using `forEach` in this case makes the code much more readable.

I intend to do some performance tuning later in the development cycle, but my initial experiments suggest that generating the TypeScript `Program` and `TypeChecker` take up about 70% of the processing time, so it is unlikely that tuning these loops will have negligible benefit to the speed of ngcc. But we will see!

mary-poppins commented:
You can preview 86732c9 at https://pr26906-86732c9.ngbuilds.io/.

mary-poppins commented:
You can preview 9ba0954 at https://pr26906-9ba0954.ngbuilds.io/.

mary-poppins commented:
You can preview 16fa15f at https://pr26906-16fa15f.ngbuilds.io/.

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -356,12 +356,7 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
-              `Unable to find matching typings (.d.ts) declaration for ${declaration.name.text} in ${declaration.getSourceFile().fileName}`);
-        }
-        return dtsDeclaration;
+        return this.dtsClassMap.get(declaration.name.text) || null;
```
Yes, being less strict here resolves an issue for me as well 👍

petebacondarwin replied:
Great! I think I am done with refactoring this. Ready for a proper review.

mary-poppins commented:
You can preview a75a204 at https://pr26906-a75a204.ngbuilds.io/.

petebacondarwin commented:
@alxhub, @JoostK & @gkalpak ready for a review. My big concern here is that there is no integration test of this functionality. I have checked that it works manually for me.

I would like to add a proper integration test, but I have been battling to move integration tests to use Bazel, and I think this is good enough to land now and then add the tests once Bazel tests are working properly.

JoostK commented on packages/compiler-cli/src/ngcc/src/analysis/private_declarations_analyzer.ts:
```diff
@@ -0,0 +1,86 @@
+    const declarations = new Map<ts.Identifier, Declaration>();
+    decorationAnalyses.forEach((compiledFile) => {
+      compiledFile.declarations.forEach(reference => {
+        const node = (reference.value as any).node;
```
This should be `WrappedNodeExpr` it seems. Could we introduce a function similar to `hasNameIdentifier` to make that a bit more explicit?

JoostK commented on packages/compiler-cli/src/ngcc/test/analysis/private_declarations_analyzer_spec.ts:
```diff
@@ -0,0 +1,142 @@
+         const analyses = analyzer.analyzeProgram(program, decorationAnalyses);
+         expect(analyses.length).toEqual(2);
+         expect(analyses).toEqual([
+           {identifier: 'ComponentB', from: '/src/b.js', dtsFrom: null},
```
Come to think of it, could we rename the symbols such that they reflect what is special about them? I had a hard time figuring out how `ComponentB` differs from `ComponentC`.

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -73,23 +74,25 @@ export class Transformer {
-        format, packageProgram, reflectionHost, isCore, r3SymbolsFile, transformDts);
-    const renderedFiles =
-        renderer.renderProgram(packageProgram, decorationAnalyses, switchMarkerAnalyses);
+        format, reflectionHost, isCore, r3SymbolsFile, entryPointFile, dtsEntryPoint);
```
The `entryPointFile` and `dtsEntryPoint` seem like symbols that could fit within the `BundleInfo` I just introduced:

https://github.com/angular/angular/pull/27055/commits/a1f12712a837dfa8daf48ab62d667ad5459381a2#diff-cfb1d3ef1d964c8d5401a9b3afeb3037R15

Let's see if that concept sticks. It would at least alleviate the chance of conflicts.

petebacondarwin replied:
I am loathe to introduce another vague term that is a collection of "stuff": `Package`, `Module`, `EntryPoint`, `Bundle` etc

This `Bundle` concept is only used for rendering, right? So perhaps we could put it all together into an `AnalyzedEntryPoint` object, which contains everything the renderer needs to do its work?

petebacondarwin replied:
But generally I like the idea of an object to hold this stuff.

petebacondarwin replied:
I have refactored the code to use a similar idea, `EntryPointBundle` after a discussion offline with @JoostK.

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/esm_renderer.ts:
```diff
@@ -27,6 +30,23 @@ export class EsmRenderer extends Renderer {
+      const fromPath = useDts ? privateExport.dtsFrom : privateExport.from;
+      if (!fromPath) {
+        throw new Error(
+            `There is no typings path for ${privateExport.identifier} in ${privateExport.from}`);
```
Can we elaborate on why having a typings file is important. Basically for the reasons outlined in the  commit message but then targeting an end user.

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/esm_renderer.ts:
```diff
@@ -27,6 +30,23 @@ export class EsmRenderer extends Renderer {
+        throw new Error(
+            `There is no typings path for ${privateExport.identifier} in ${privateExport.from}`);
+      }
+      const basePath = fromPath.replace(/\.(js|d\.ts$)/, '');
```
Should include `$` for js as well:

```suggestion
      const basePath = fromPath.replace(/\.(js|d\.ts)$/, '');
```

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -124,10 +137,18 @@ export abstract class Renderer {
 
+    // Add exports to the entry-point file
+    if (sourceFile === this.entryPoint) {
+      const entryPointBasePath = this.entryPoint.fileName.replace(/\.(js|d\.ts$)/, '');
```
```suggestion
      const entryPointBasePath = this.entryPoint.fileName.replace(/\.(js|d\.ts)$/, '');
```

JoostK reviewed (commented):
I should have seen this PR earlier, as I should have used it as base commit (this badly conflicts with #27055, 😭)

mary-poppins commented:
You can preview ef5fafc at https://pr26906-ef5fafc.ngbuilds.io/.

mary-poppins commented:
You can preview c9e8d33 at https://pr26906-c9e8d33.ngbuilds.io/.

petebacondarwin committed:
refactor(ivy): ngcc - make `EntryPoint` an interface rather than a type

By inverting the relationship between `EntryPointPaths` and
`EntryPointFormat` we can have interfaces rather than types.

Thanks to @gkalpak for this idea.

petebacondarwin committed:
feat(ivy): register references from NgModule annotations

The `NgModuleDecoratorHandler` can now register all the references that
it finds in the `NgModule` metadata, such as `declarations`, `imports`,
`exports` etc.

This information can then be used by ngcc to work out if any of these
references are internal only and need to be manually exported from a
library's entry-point.

petebacondarwin committed:
style(ivy): ngcc - fix misspelled method

mary-poppins commented:
You can preview 3817acb at https://pr26906-3817acb.ngbuilds.io/.

mary-poppins commented:
You can preview cfed4e1 at https://pr26906-cfed4e1.ngbuilds.io/.

mary-poppins commented:
You can preview 1957035 at https://pr26906-1957035.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -78,21 +79,19 @@ export class DecorationAnalyzer {
   analyzeProgram(program: ts.Program): DecorationAnalyses {
     const decorationAnalyses = new DecorationAnalyses();
-    const analysedFiles =
+    const analyzedFiles =
```
Just trying to be consistent :-)

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1151,3 +1152,22 @@ function getFarLeftIdentifier(propertyAccess: ts.PropertyAccessExpression): ts.I
+      }
+      const declaration = exportedSymbol.declarations[0];
+      const name = exportedSymbol.name;
+      if (declaration && ts.isClassDeclaration(declaration) && !dtsClassMap.has(name)) {
```
This can happen because we are not only looking at exports in `srcFile` but also exports from other source files too. `collectExportedClasses` gets called for each source file in the program, and each one adds to the `dtsMap`.

It is true that only one could ever be exported publicly from the entry-point, but this PR has to deal with "internal" classes too.

alxhub replied:
I think we should still do something other than silently ignore when there's more than one class with the same name. Can we store an array of them, or put a special marker in the `Map` that will cause an error if we try to publicly export one of these?

petebacondarwin replied:
There are two scenarios, we should deal with:

1) There is both a public and one or more internal exports with the same name.
2) There is no public but more than one internal export with the same name.

In most cases it is the public one we are after, right? So should we just ignore additional internal matching names?

But if we are trying to create an export for an internal class and we have multiple classes with the same name, then I think this should probably error, since we don't know which we are supposed to be exporting.

petebacondarwin replied:
So I think that the best solution is for `getDtsDeclarationOfClass` to return an array of possible declarations, with the public one (if any) first. The user of `getDtsDeclarationOfClass` then has to decide whether having more than one result is an error or not.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -121,7 +120,13 @@ export class DecorationAnalyzer {
+    const declarations: R3Reference[] = [];
+    compiledClasses.forEach(c => {
+      if (c.handler instanceof NgModuleDecoratorHandler) {
+        declarations.push(...c.analysis.ngModuleDef.declarations);
```
OK, I guess you mean `DecoratorHandler` there?

petebacondarwin replied:
I just realised that we probably ought to check the `imports`, `exports` and `bootstrap` collections too, right? You could have a scenario where an `NgModule` references another `NgModule`, as an import say, but that second `NgModule` is not publicly exported... right?

petebacondarwin committed:
feat(ivy): ngcc - add `PrivateDeclarationsAnalyzer`

This analyzer searches the source for declared classes that are not
exported publicly from the entry-point.

mary-poppins commented:
You can preview 224b691 at https://pr26906-224b691.ngbuilds.io/.

petebacondarwin commented:
@alxhub and @JoostK I have completely refactored the private exports bit of this PR. I also moved all the refactoring of entry-point bundles to the start of the PR so that it is not dependent upon the private export stuff. This means that it would be reusable if we decided not to take the last few commits.

PTAL!

mary-poppins commented:
You can preview 6920a45 at https://pr26906-6920a45.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1151,3 +1152,28 @@ function getFarLeftIdentifier(propertyAccess: ts.PropertyAccessExpression): ts.I
+      if (exportedSymbol.flags & ts.SymbolFlags.Alias) {
+        exportedSymbol = checker.getAliasedSymbol(exportedSymbol);
+      }
+      const declaration = exportedSymbol.declarations[0];
```
I'm not sure about the ordering of `declarations` here -- we could probably check `.valueDeclaration` and not `.declarations[0]`.

petebacondarwin replied:
I think there was a reason why I used `declarations[0]` rather than `valueDeclaration` but I can't remember now. I'll take a look and check.

petebacondarwin replied:
I switched it and no tests fail, so... let's go with `valueDeclaration` until something breaks.
(I agree that we are only interested in concrete classes here, so they should have a value declaration).

alxhub commented on packages/compiler-cli/src/ngcc/src/packages/entry_point.ts:
```diff
@@ -11,22 +11,26 @@ import * as fs from 'fs';
-export type EntryPointPaths = {
-  [Format in EntryPointFormat]?: string;
-};
+export type EntryPointFormat = keyof(EntryPointPaths);
```
I don't mind either way, but the parentheses here are optional, `keyof EntryPointPaths` works just as well.

In general this change appears to be a nice simplification! 🥂

alxhub commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -69,25 +69,26 @@ export class Transformer {
     const rootDirs = this.getRootDirs(host, options);
-    const isCore = entryPoint.name === '@angular/core';
-    const r3SymbolsPath = isCore ? this.findR3SymbolsPath(dirname(entryPointFilePath)) : null;
+    const r3SymbolsPath =
```
Help me understand a little here - what happens if the current entypoint is a FESM file where there's no individual `r3_symbols` file to import from?

JoostK replied:
@alxhub for FESMs it works for the same reason as outlined in #27055 (no imports will be generated at all when compiling core). The ESM bundles are currently broken however (they reference `@angular/core` imports from within core itself); that is actually being fixed in #27055.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/references_registry.ts:
```diff
@@ -0,0 +1,61 @@
+   */
+  getDeclarationMap(): Map<ts.Identifier, Declaration> {
+    // We lazily create the map when asked for so that it doesn't slow down
+    // ngtsc compilation.
```
This comment could use some more context, specifically the mention of ngcc that is currently the only consumer of this logic. It's now confusing that `ngtsc` contains code that it apparently doesn't use itself.

Thinking about this more, we might want to convert `ReferencesRegistry` into an interface and pass an `IntoTheVoidRegistry` 😉  with ngtsc, and move the current impl to ngcc such that this logic becomes ngcc specific. Doing so would also justify this implementation's choice to only store `ResolvedReference`s and ignore all others.

petebacondarwin replied:
Yes I like that idea.

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/esm_renderer.ts:
```diff
@@ -26,6 +29,26 @@ export class EsmRenderer extends Renderer {
+  addExports(
+      output: MagicString, entryPointBasePath: string, privateExports: ExportInfo[],
+      useDts: boolean): void {
+    privateExports.forEach(privateExport => {
```
This `addExports` now has a much different signature compared to `addImports`. I think the logic that is now in this method should be computed by the caller, similar to how `addImports` only gets the import data.

petebacondarwin replied:
Ahah! I just realised that I did not address this @JoostK ... I'll take a look tonight

petebacondarwin replied:
I simplified it a bit.

petebacondarwin committed:
fix(ivy): ngcc - support typings (d.ts) classes that are not publicly exported

If a decorated class is not publicly exported via an entry-point then the
previous approach to finding the associated typings file failed.

Now we ensure that we extract all the class declarations from the
dtsTypings program, even if they are not exported from the entry-point.
This is achieved by also parsing statements of each source file, rather
than just parsing classes that are exported from the entry-point.

Because we now look at all the files, it is possible for there to be multiple
class declarations with the same local name. In this case, only the first
declaration with a given name is added to the map; subsequent classes are
ignored.

We are most interested in classes that are publicly exported from the
entry-point, so these are added to the map first, to ensure that they are
not ignored.

mary-poppins commented:
You can preview 6ef8be2 at https://pr26906-6ef8be2.ngbuilds.io/.

petebacondarwin commented:
The current rebase broke the fix from @JoostK for r3 symbols in .d.ts files. I have it fixed locally but I need to rework the changes back into the commits - again!

mary-poppins commented:
You can preview 3aed964 at https://pr26906-3aed964.ngbuilds.io/.

petebacondarwin committed:
style(ivy): ngcc - fix typos

petebacondarwin committed:
refactor(ivy): ngcc - remove unused code

petebacondarwin committed:
refactor(ivy): ngcc - subclasses do not need to declare `protected` variables

These variables are already declared as protected in the super class, and
so it is redundant to do it again in the subclasses.

petebacondarwin committed:
refactor(ivy): ngcc - encapsulate variables into "bundles"

There are a number of variables that need to be passed around
the program, in particular to the renderers, which benefit from being
stored in well defined objects.

The new `EntryPointBundle` structure is a specific format of an entry-point
and contains the compiled `BundleProgram` objects for the source and typings,
if appropriate.

This change helps with future refactoring, where we may need to add new
properties to this object. It allows us to maintain more stable APIs between
the constituent parts of ngcc, rather than passing lots of primitive values
around throughout the program.

petebacondarwin committed:
feat(ivy): ngcc - render private declaration exports

Ngcc will now render additional exports for classes that are referenced in
`NgModule` decorated classes, but which were not publicly exported
from an entry-point of the package.

This is important because when ngtsc compiles libraries processed by ngcc
it needs to be able to publcly access decorated classes that are referenced
by `NgModule` decorated classes in order to build templates that use these
classes.

Doing this re-exporting is not without its risks. There are chances that
the class is not exported correctly: there may already be similarly named
exports from the entry-point or the class may be being aliased. But there
is not much more we can do from the point of view of ngcc to workaround
such scenarios. Generally, packages should have been built so that this
approach works.

mary-poppins commented:
You can preview 6f21d4e at https://pr26906-6f21d4e.ngbuilds.io/.

mary-poppins commented:
You can preview 1635469 at https://pr26906-1635469.ngbuilds.io/.

petebacondarwin committed:
test(ivy): ngcc - update integration test dependency lock file

mary-poppins commented:
You can preview e10bce0 at https://pr26906-e10bce0.ngbuilds.io/.

mary-poppins commented:
You can preview 59f21c9 at https://pr26906-59f21c9.ngbuilds.io/.

mary-poppins commented:
You can preview f044bc7 at https://pr26906-f044bc7.ngbuilds.io/.

mary-poppins commented:
You can preview 3419c2b at https://pr26906-3419c2b.ngbuilds.io/.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/references_registry.ts:
```diff
@@ -0,0 +1,39 @@
+   * Create and return a mapping for the registered resolved references.
+   * @returns A map of reference identifiers to reference declarations.
+   */
+  getDeclarationMap(): Map<ts.Identifier, Declaration>;
```
I think this accessor can be removed from the interface, as it's only consumed by ngcc and perhaps specific to ngcc as well. The `Registry` part of the name would then become inaccurate though, so perhaps this is just fine here.

petebacondarwin replied:
I'll see what @alxhub says but my inclination would be to leave it in place for now.

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/entry_point_bundle.ts:
```diff
@@ -0,0 +1,58 @@
+  format: EntryPointFormat;
+  isFlat: boolean;
+  rootDirs: string[];
+  src: BundleProgram;
```
I like this, makes sense and achieves quite a cleanup in the transformer 👍

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/esm5_renderer.ts:
```diff
@@ -9,14 +9,14 @@ import * as ts from 'typescript';
-      host: NgccReflectionHost, bundle: BundleInfo, sourcePath: string, targetPath: string,
-      transformDts: boolean) {
-    super(host, bundle, sourcePath, targetPath, transformDts);
+      protected host: NgccReflectionHost, isCore: boolean, bundle: EntryPointBundle,
```
Looks like the `protected` can be dropped

JoostK commented on packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts:
```diff
@@ -227,16 +236,38 @@ A.ngDirectiveDef = ɵngcc0.ɵdefineDirective({ type: A, selectors: [["", "a", ""
+        const typingsFile = result.find(f => f.path === '/typings/file.d.ts') !;
+        expect(typingsFile.contents)
+            .toContain(
+                'foo(x: number): number;\n    static ngDirectiveDef: ɵngcc0.ɵDirectiveDefWithMeta');
```
I realize that there's no such test for `@angular/core`, in which we expect `ngDirectiveDef: DirectiveDefWithMeta`. I did add that assertion to the integration test but might as well test it from here.

JoostK replied:
Sure!

JoostK reviewed (commented):
Looks great!

<sup>PS. Where did my code go? 😝</sup>

petebacondarwin committed:
fixup! refactor(ivy): ngcc - encapsulate variables into "bundles"

petebacondarwin committed:
fixup! refactor(ivy): ngcc - encapsulate variables into "bundles"

mary-poppins commented:
You can preview c1139c0 at https://pr26906-c1139c0.ngbuilds.io/.

mhevery commented:
http://test/OCL:223178683:BASE:223178801:1543425187595:be63129f

petebacondarwin commented:
Caretaker: can you check the status of the presubmit and then merge if all is good?

IgorMinar commented:
OCL:223178683:BASE:223462597:1543556237402:8cd641d9

IgorMinar commented:
http://test/OCL:223563071:BASE:223563100:1543612151002:79d21975
